src/shim/runtime/v2/rune/v2/agent.go

```diff
@@ -0,0 +1,279 @@
+		c.dead = true
+		return err
+	}
+
+	c.installReqFunc(client)
+	c.client = client
+
+	return nil
+}
+
+func (c *agent) disconnect(ctx context.Context) error {
+	c.Lock()
+	defer c.Unlock()
+
+	if c.client == nil {
+		return nil
+	}
+
+	if err := c.client.Close(); err != nil && grpcStatus.Convert(err).Code() != codes.Canceled {
+		return err
+	}
+
+	c.client = nil
+	c.reqHandlers = nil
+
+	return nil
+}
+
+type reqFunc func(context.Context, interface{}) (interface{}, error)
+
+func (c *agent) installReqFunc(client *agentClient.AgentClient) {
+	c.reqHandlers = make(map[string]reqFunc)
+	c.reqHandlers[grpcPullImageRequest] = func(ctx context.Context, req interface{}) (interface{}, error) {
+		return c.client.ImageServiceClient.PullImage(ctx, req.(*grpc.PullImageRequest))
+	}
+}
+
+func (c *agent) sendReq(spanCtx context.Context, request interface{}) (interface{}, error) {
+	if err := c.connect(spanCtx); err != nil {
+		return nil, err
+	}
+
+	defer c.disconnect(spanCtx)
+
+	msgName := proto.MessageName(request.(proto.Message))
+
+	c.Lock()
+
+	if c.reqHandlers == nil {
+		return nil, errors.New("Client has already disconnected")
+	}
+
+	handler := c.reqHandlers[msgName]
+	if msgName == "" || handler == nil {
+		return nil, errors.New("Invalid request type")
+	}
+
+	c.Unlock()
+
+	message := request.(proto.Message)
+	ctx, cancel := context.WithTimeout(spanCtx, defaultRequestTimeout)
+	if cancel != nil {
+		defer cancel()
+	}
+	c.Logger().WithField("name", msgName).WithField("req", message.String()).Trace("sending request")
+
+	return handler(ctx, request)
+}
+
+func (c *agent) PullImage(ctx context.Context, req *image.PullImageReq) (*image.PullImageResp, error) {
+	cid, err := getContainerID(req.Image)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create dir for store unionfs image (based on sefs)
+	sefsDir := filepath.Join(c.Bundle, "rootfs/images/", cid, "sefs")
+	lowerDir := filepath.Join(sefsDir, "lower")
+	upperDir := filepath.Join(sefsDir, "upper")
+	for _, dir := range []string{lowerDir, upperDir} {
+		if err := os.MkdirAll(dir, 0700); err != nil && !os.IsExist(err) {
+			return nil, err
+		}
+	}
+
+	r := &grpc.PullImageRequest{
+		Image:       req.Image,
+		ContainerId: cid,
+	}
+	resp, err := c.sendReq(ctx, r)
+	if err != nil {
+		c.Logger().Errorf("agent enclave container pull image err. %v", err)
+		return nil, err
+	}
+	response := resp.(*grpc.PullImageResponse)
+	return &image.PullImageResp{
+		ImageRef: response.ImageRef,
+	}, nil
+}
+
+// The function creates agent enclave container based on a pre-installed OCI bundle
+func createAgentContainer(ctx context.Context, s *service, r *taskAPI.CreateTaskRequest) (*runc.Container, error) {
+	dir := filepath.Join(agentContainerRootDir, r.ID)
+	upperDir := path.Join(dir, "upper")
+	workDir := path.Join(dir, "work")
+	destDir := path.Join(dir, "merged")
+	for _, dir := range []string{upperDir, workDir, destDir} {
+		if err := os.MkdirAll(dir, 0700); err != nil && !os.IsExist(err) {
```

Comment:
There are multiple instances of `0700` on this PR so I'd consider creating a global `const` for it.